proc/threads.go

```diff
@@ -145,73 +146,167 @@ func topframe(g *G, thread *Thread) (Stackframe, error) {
+				deferreturns = append(deferreturns, instr.Loc.PC)
+			}
 		}
-	}
 
-	if !covered {
-		fn := dbp.goSymTable.PCToFunc(topframe.Ret)
-		if g != nil && fn != nil && fn.Name == "runtime.goexit" {
-			return nil
+		// Set breakpoint on the most recently deferred function (if any)
+		var deferpc uint64 = 0
+		if dbp.SelectedGoroutine != nil && dbp.SelectedGoroutine.DeferPC != 0 {
+			_, _, deferfn := dbp.goSymTable.PCToLine(dbp.SelectedGoroutine.DeferPC)
+			var err error
+			deferpc, err = dbp.FirstPCAfterPrologue(deferfn, false)
+			if err != nil {
+				return err
+			}
+		}
+		if deferpc != 0 && deferpc != topframe.Current.PC {
+			bp, err := dbp.SetTempBreakpoint(deferpc, NextBreakpoint, cond)
+			if err != nil {
+				if _, ok := err.(BreakpointExistsError); !ok {
+					dbp.ClearTempBreakpoints()
+					return err
+				}
+			}
+			if bp != nil {
+				bp.DeferCond = true
+				bp.DeferReturns = deferreturns
+			}
 		}
 	}
-	if deferpc != 0 {
-		pcs = append(pcs, deferpc)
+
+	// Add breakpoints on all the lines in the current function
+	pcs := dbp.lineInfo.AllPCsBetween(topframe.FDE.Begin(), topframe.FDE.End()-1, topframe.Current.File)
+
+	if !csource {
+		var covered bool
+		for i := range pcs {
+			if topframe.FDE.Cover(pcs[i]) {
+				covered = true
+				break
+			}
+		}
+
+		if !covered {
+			fn := dbp.goSymTable.PCToFunc(topframe.Ret)
+			if dbp.SelectedGoroutine != nil && fn != nil && fn.Name == "runtime.goexit" {
+				return nil
+			}
+		}
 	}
+
+	// Add a breakpoint on the return address for the current frame
 	pcs = append(pcs, topframe.Ret)
-	return dbp.setTempBreakpoints(topframe.Current.PC, pcs, sameGoroutineCondition(dbp.SelectedGoroutine))
+	return dbp.setTempBreakpoints(topframe.Current.PC, pcs, NextBreakpoint, cond)
 }
 
-// Set a breakpoint at every reachable location, as well as the return address. Without
-// the benefit of an AST we can't be sure we're not at a branching statement and thus
-// cannot accurately predict where we may end up.
-func (dbp *Process) cnext(topframe Stackframe) error {
-	pcs := dbp.lineInfo.AllPCsBetween(topframe.FDE.Begin(), topframe.FDE.End(), topframe.Current.File)
-	pcs = append(pcs, topframe.Ret)
-	return dbp.setTempBreakpoints(topframe.Current.PC, pcs, sameGoroutineCondition(dbp.SelectedGoroutine))
+func (dbp *Process) setStepIntoBreakpoint(text []AsmInstruction, cond ast.Expr) error {
+	if len(text) <= 0 {
+		return nil
+	}
+
+	instr := text[0]
+
+	if instr.DestLoc == nil || instr.DestLoc.Fn == nil {
+		return nil
+	}
+
+	fn := instr.DestLoc.Fn
+
+	// Ensure PC and Entry match, otherwise StepInto is likely to set
+	// its breakpoint before DestLoc.PC and hence run too far ahead.
+	// Calls to runtime.duffzero and duffcopy have this problem.
+	if fn.Entry != instr.DestLoc.PC {
+		return nil
+	}
+
+	// Skip unexported runtime functions
+	if strings.HasPrefix(fn.Name, "runtime.") && !isExportedRuntime(fn.Name) {
```

Comment:
If I omit the `strings.HasPrefix(fn.Name, "runtime.")` we hide everything except exported runtime functions, what we want is hide all runtime functions except the ones that are exported.